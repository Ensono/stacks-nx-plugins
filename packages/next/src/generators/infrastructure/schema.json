{
    "$schema": "http://json-schema.org/schema",
    "cli": "nx",
    "$id": "Next",
    "title": "Ensono Stacks Next Infrastructure Generator",
    "description": "The @ensono-stacks/next:infrastructure will add deployment infrastructure to an existing next project. Including generating a docker file, Heml chart and Terraform config",
    "type": "object",
    "properties": {
        "project": {
            "type": "string",
            "description": "The name of the project.",
            "alias": "p",
            "$default": {
                "$source": "projectName"
            },
            "x-prompt": "What is the name of the project for this component?"
        },
        "openTelemetry": {
            "type": "boolean",
            "description": "Auto-instrumentation for opentelemetry",
            "alias": "o",
            "$default": "false",
            "x-prompt": "Would you like auto-instrumentation for OpenTelemetry?"
        }
    },
    "required": ["project"]
}
