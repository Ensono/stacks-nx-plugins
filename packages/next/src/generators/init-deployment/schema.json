{
    "$schema": "http://json-schema.org/schema",
    "cli": "nx",
    "$id": "InitDeployment",
    "title": "Ensono Stacks Next Init Deployment Generator",
    "description": "The @ensono-stacks/next:init-deployment will add deployment infrastructure to an existing next project. Including generating a docker file, Heml chart and Terraform config",
    "type": "object",
    "properties": {
        "project": {
            "type": "string",
            "description": "The name of the project.",
            "alias": "p",
            "$default": {
                "$source": "projectName"
            },
            "x-prompt": "What is the name of the project for this component?"
        },
        "libraryName": {
            "type": "string",
            "description": "Name provided for generated helm chart library",
            "alias": "l",
            "default": "stacks-helm-chart",
            "x-prompt": "Would you like to specify a name for the generated helm chart library?"
        }
    },
    "required": ["project"]
}
