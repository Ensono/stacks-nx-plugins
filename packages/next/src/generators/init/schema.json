{
    "$schema": "http://json-schema.org/schema",
    "cli": "nx",
    "$id": "Next",
    "title": "",
    "type": "object",
    "properties": {
        "project": {
            "type": "string",
            "description": "The name of the project.",
            "alias": "p",
            "$default": {
                "$source": "projectName"
            },
            "x-prompt": "What is the name of the project for this component?"
        }
    },
    "required": ["project"]
}
