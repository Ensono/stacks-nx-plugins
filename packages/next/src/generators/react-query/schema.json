{
    "$schema": "http://json-schema.org/schema",
    "$id": "ReactQuery",
    "title": "Ensono Stacks React Query Generator",
    "description": "The @ensono-stacks/next:react-query generator adds React Query to the Next App",
    "type": "object",
    "properties": {
        "project": {
            "type": "string",
            "description": "The name of the project.",
            "alias": "p",
            "$default": {
                "$source": "projectName"
            },
            "x-prompt": "What is the name of the project?",
            "x-priority": "important"
        },
        "skipPackageJson": {
            "type": "boolean",
            "default": false,
            "description": "Do not add dependencies to `package.json`."
        }
    },
    "required": [
        "project"
    ]
}
