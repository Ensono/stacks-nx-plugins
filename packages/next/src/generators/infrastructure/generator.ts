import { StacksConfigError } from '@ensono-stacks/core';
import {
    joinPathFragments,
    readProjectConfiguration,
    GeneratorCallback,
    Tree,
} from '@nrwl/devkit';
import { runTasksInSerial } from '@nrwl/workspace/src/utilities/run-tasks-in-serial';
import chalk from 'chalk';

import { NextGeneratorSchema } from './schema';
import { addInfrastructure } from './utils/add-infrastructure';
import { formatFiles } from './utils/format-files';

export default async function infrastructureGenerator(
    tree: Tree,
    options: NextGeneratorSchema,
) {
    const tasks: GeneratorCallback[] = [];
    const project = readProjectConfiguration(tree, options.project);

    try {
        tasks.push(...addInfrastructure(tree, project));
    } catch (error) {
        if (error instanceof StacksConfigError) {
            console.warn(chalk.yellow`Missing Stacks configuration in nx.json`);
            console.log(
                '  Infrastructure code will not be applied now, but you can rerun this task when you update the configuration.',
            );
            console.log(
                '    nx g @ensono-stacks/next:infrastructure --project [project_name]',
            );
            return () => {};
        }

        throw error;
    }

    // exclude helm yaml files from initial format when generating the files
    await formatFiles(tree, [
        joinPathFragments(project.root, 'build', 'helm', '**', '*.yaml'),
    ]);

    return runTasksInSerial(...tasks);
}
