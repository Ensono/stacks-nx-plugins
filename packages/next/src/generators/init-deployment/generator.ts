import {
    formatFiles,
    executedDependantGenerator,
    hasGeneratorExecutedForProject,
    StacksError,
} from '@ensono-stacks/core';
import {
    joinPathFragments,
    readProjectConfiguration,
    GeneratorCallback,
    Tree,
} from '@nrwl/devkit';
import { runTasksInSerial } from '@nrwl/workspace/src/utilities/run-tasks-in-serial';
import chalk from 'chalk';

import { NextGeneratorSchema } from './schema';
import { addInfrastructure } from './utils/add-infrastructure';

export default async function initDeploymentGenerator(
    tree: Tree,
    options: NextGeneratorSchema,
) {
    if (
        !executedDependantGenerator(
            tree,
            'WorkspaceDeployment',
            options.project,
        )
    )
        return false;
    if (
        hasGeneratorExecutedForProject(
            tree,
            options.project,
            'NextInitDeployment',
            true,
        )
    )
        return false;

    const tasks: GeneratorCallback[] = [];
    const project = readProjectConfiguration(tree, options.project);

    try {
        tasks.push(...addInfrastructure(tree, project, options));
    } catch (error) {
        if (error instanceof StacksError) {
            console.warn(chalk.yellow`Missing Stacks configuration in nx.json`);
            console.log(
                '  Infrastructure code will not be applied now, but you can rerun this task when you update the configuration.',
            );
            console.log(
                '    nx g @ensono-stacks/next:init-deployment --project [project_name]',
            );
            return () => {};
        }

        throw error;
    }

    // exclude helm yaml files from initial format when generating the files
    await formatFiles(tree, [
        joinPathFragments(
            'libs',
            options.libraryName,
            'build',
            'helm',
            '**',
            '*.yaml',
        ),
    ]);

    return runTasksInSerial(...tasks);
}
