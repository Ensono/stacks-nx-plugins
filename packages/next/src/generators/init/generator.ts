import {
    formatFiles,
    formatFilesWithEslint,
    addCustomTestConfig,
    deploymentGeneratorMessage,
    hasGeneratorExecutedForProject,
    verifyPluginCanBeInstalled,
    tsMorphTree,
} from '@ensono-stacks/core';
import {
    GeneratorCallback,
    joinPathFragments,
    readProjectConfiguration,
    Tree,
    updateProjectConfiguration,
} from '@nx/devkit';
import { runTasksInSerial } from '@nx/workspace/src/utilities/run-tasks-in-serial';
import chalk from 'chalk';
import path from 'path';

import { NextGeneratorSchema } from './schema';
import { addEslint } from './utils/eslint';
import { eslintFix } from './utils/eslint-fix';
import {
    addReactAxeConfigToApp,
    addReactAxeDependency,
} from './utils/react-axe';
import { addStorybook } from './utils/storybook';
import updateTsConfig from './utils/tsconfig';

export default async function initGenerator(
    tree: Tree,
    options: NextGeneratorSchema,
) {
    verifyPluginCanBeInstalled(tree, options.project);

    if (hasGeneratorExecutedForProject(tree, options.project, 'NextInit'))
        return false;

    const tasks: GeneratorCallback[] = [];
    const project = readProjectConfiguration(tree, options.project);

    const update = { ...project };

    if (
        project.sourceRoot &&
        project.sourceRoot ===
            update.targets?.build?.configurations?.development?.outputPath
    ) {
        update.targets.build.configurations.development = {};
    }

    updateProjectConfiguration(tree, project.name, update);

    tasks.push(addEslint(tree, project.root), addStorybook(tree, options));

    const ciCoverageConfig = {
        ci: {
            collectCoverage: true,
            coverageReporters: ['text', 'html'],
            collectCoverageFrom: [
                './**/*.{js,jsx,ts,tsx}',
                './!**/.next/**',
                './!**/*.d.ts',
                './!**/*.config.*',
                './!**/_app.*',
            ],
            codeCoverage: true,
            ci: true,
        },
    };

    await addCustomTestConfig(
        tree,
        readProjectConfiguration(tree, options.project),
        project.name,
        ciCoverageConfig,
    );

    tasks.push(
        formatFilesWithEslint(options.project),
        addReactAxeDependency(tree),
    );

    // update tsconfig.json
    updateTsConfig(
        tree,
        project,
        path.join(project.sourceRoot, 'tsconfig.json'),
        ['next.config.js'],
    );

    // update tsconfig.spec.json
    updateTsConfig(
        tree,
        project,
        path.join(project.sourceRoot, 'tsconfig.spec.json'),
    );

    const morphTree = tsMorphTree(tree);

    console.info(chalk.green`Attempting to add configuration for react-axe`);
    addReactAxeConfigToApp(project, morphTree);
    console.info(
        chalk.green`continuing execution of next:init generator after attempting to add react-axe`,
    );
    eslintFix(project, tree);

    // exclude helm yaml files from initial format when generating the files
    await formatFiles(tree, [
        joinPathFragments(project.root, 'build', 'helm', '**', '*.yaml'),
    ]);

    return runTasksInSerial(...tasks, () =>
        deploymentGeneratorMessage(
            tree,
            `nx g @ensono-stacks/next:init-deployment --project ${options.project}`,
        ),
    );
}
