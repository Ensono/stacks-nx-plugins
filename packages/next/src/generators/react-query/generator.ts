import {
    formatFiles,
    formatFilesWithEslint,
    hasGeneratorExecutedForProject,
    tsMorphTree,
    verifyPluginCanBeInstalled,
} from '@ensono-stacks/core';
import {
    joinPathFragments,
    logger,
    readProjectConfiguration,
    Tree,
} from '@nx/devkit';
import { runTasksInSerial } from '@nx/workspace/src/utilities/run-tasks-in-serial';

import { ReactQueryGeneratorSchema } from './schema';
import { installDependencies } from './utils/dependancies';
import { updateESLint } from './utils/eslint';
import { addQueryClientProviderToApp } from './utils/query-client-provider';

export async function reactQueryGenerator(
    tree: Tree,
    options: ReactQueryGeneratorSchema,
) {
    verifyPluginCanBeInstalled(tree, options.project);

    if (
        hasGeneratorExecutedForProject(
            tree,
            options.project,
            'ReactQuery',
            true,
        )
    )
        return false;

    const project = readProjectConfiguration(tree, options.project);

    const morphTree = tsMorphTree(tree);

    addQueryClientProviderToApp(project, morphTree);

    updateESLint(tree, project.root);

    await formatFiles(tree, [
        joinPathFragments(project.root, 'build', 'helm', '**', '*.yaml'),
    ]);

    return runTasksInSerial(
        options.skipPackageJson ? () => {} : installDependencies(tree),
        formatFilesWithEslint(options.project),
        () => {
            logger.warn(
                `React Query has been added to your app's _app.tsx file`,
            );
        },
    );
}

export default reactQueryGenerator;
