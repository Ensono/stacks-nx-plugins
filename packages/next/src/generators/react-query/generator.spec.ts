import { addStacksAttributes } from '@ensono-stacks/test';
import { Tree, readJson } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { applicationGenerator } from '@nx/next';

import generator from './generator';
import { ReactQueryGeneratorSchema } from './schema';
import { nextAppWithProviders } from './test/fixtures';
import { nextAppWithDestructuredProperties } from '../next-auth/test/fixtures';

describe('react-query generator', () => {
    let appTree: Tree;
    const options: ReactQueryGeneratorSchema = {
        project: 'next-app',
    };

    beforeEach(async () => {
        appTree = createTreeWithEmptyWorkspace();
        await applicationGenerator(appTree, {
            name: 'next-app',
            style: 'css',
            appDir: false,
        });

        addStacksAttributes(appTree, options.project);
    });

    describe('executedGenerators', () => {
        beforeEach(async () => {
            await generator(appTree, { ...options });
        });

        it('should update nx.json and tag executed generator true', async () => {
            const nxJson = readJson(appTree, 'nx.json');

            expect(
                nxJson.stacks.executedGenerators.project[
                    options.project
                ].includes('ReactQuery'),
            ).toBe(true);
        });

        it('should return false from method and exit generator if already executed', async () => {
            await generator(appTree, {
                ...options,
            });

            const eslintConfig = readJson(appTree, '.eslintrc.json');

            expect(Object.keys(eslintConfig.extends)).toEqual(
                expect.arrayContaining([
                    'plugin:@tanstack/eslint-plugin-query/recommended',
                ]),
            );
        });

        it('should eslint config with react query plugin', async () => {
            const gen = await generator(appTree, {
                ...options,
            });

            expect(gen).toBe(false);
        });
    });

    it('should configure app if there are already wrapping react providers', async () => {
        appTree.write('next-app/pages/_app.tsx', nextAppWithProviders);
        await generator(appTree, options);

        const AppTsx = appTree.read('next-app/pages/_app.tsx');

        expect(AppTsx.toString()).toMatchSnapshot();
    });

    it('should configure app if pageProps is already destructured', async () => {
        appTree.write(
            'next-app/pages/_app.tsx',
            nextAppWithDestructuredProperties,
        );
        await generator(appTree, options);

        const AppTsx = appTree.read('next-app/pages/_app.tsx');

        expect(AppTsx.toString()).toMatchSnapshot();
    });

    it('should skip installing dependencies', async () => {
        await generator(appTree, { ...options, skipPackageJson: true });

        const packageJson = readJson(appTree, 'package.json');

        expect(Object.keys(packageJson.dependencies)).not.toEqual(
            expect.arrayContaining(['react-query']),
        );
    });
});
