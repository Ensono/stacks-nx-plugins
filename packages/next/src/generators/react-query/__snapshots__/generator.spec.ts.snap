// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-query generator should configure app if there are already wrapping react providers 1`] = `
"import { SessionProvider } from 'next-auth/react';
function SomeProvider({ children }: PropsWithChildren) {
    return <>{children}</>;
}
function CustomApp({
    Component,
    pageProps: { session, ...pageProps },
}: AppProps) {
    return (
        <SessionProvider session={session}>
            <SomeProvider>
                <Component {...pageProps} />
            </SomeProvider>
        </SessionProvider>
    );
}
export default CustomApp;
"
`;